contracts/exchange/test/transactions_unit_tests.ts

```diff
@@ -0,0 +1,694 @@
+            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(
+                transactionHash,
+                executableError.encode(),
+            );
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction, randomSignature(), {
+                from: accounts[1],
+            });
+            return expect(tx).to.revertWith(expectedError);
+        });
+
+        it('should revert if the signer != msg.sender and the signature is valid but the delegatecall fails', async () => {
+            // This calldata is encoded to fail when it hits the executable function.
+            const callData = getExecutableCallData(false, constants.NULL_BYTES, constants.NULL_BYTES);
+            const currentTimestamp = await getLatestBlockTimestampAsync();
+            const transaction = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
+                signerAddress: accounts[1], // This is different than the account that will be used to send.
+                data: callData,
+                domain,
+            };
+            const validSignature = randomSignature(); // Valid because length != 2
+            const transactionHash = transactionHashUtils.getTransactionHashHex(transaction);
+            const executableError = new StringRevertError('EXECUTABLE_FAILED');
+            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(
+                transactionHash,
+                executableError.encode(),
+            );
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction, validSignature, {
+                from: accounts[0],
+            });
+            return expect(tx).to.revertWith(expectedError);
+        });
+
+        it('should succeed with the correct return hash and event emitted', async () => {
+            // This calldata is encoded to succeed when it hits the executable function.
+            const callData = getExecutableCallData(true, constants.NULL_BYTES, '0xdeadbeef');
+            const currentTimestamp = await getLatestBlockTimestampAsync();
+            const validSignature = randomSignature(); // Valid because length != 2
+            const transaction = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
+                signerAddress: accounts[1], // This is different than the account that will be used to send.
+                data: callData,
+                domain,
+            };
+            const transactionHash = transactionHashUtils.getTransactionHashHex(transaction);
+
+            // Verify that the returndata of the transaction is 0xDEADBEEF
+            const result = await transactionsContract.executeTransaction.callAsync(transaction, validSignature, {
```

Comment:
The `TransactionHelper` class in `@0x/contracts-test-utils` might reduce some of the code in blocks where you do a `callAsync`, `sendTransactionAsync`, and `getTxWithDecodedLogsAsync` since it does all three.